File: server/src/app/rooms.py
from flask import Blueprint, jsonify, g, request
from pydantic import ValidationError

from src.logging import get_logger
from src.db import Database
from src.models import Room

from .middleware import data_has, ensure_not_none, load_room


logger = get_logger(__name__)
rooms_bp = Blueprint('rooms', __name__)


rooms_bp.before_request(load_room)


@rooms_bp.get('/')
def list_rooms():
    logger.info('processing request to list rooms')
    db: Database = g.db
    rooms = db.rooms.find_many()
    room_data = [room.model_dump() for room in rooms or []]
    return jsonify({"data": {"rooms": room_data}})


@rooms_bp.post('/')
@data_has('name', optional=True)
def create_room():
    logger.info('processing create room request', {'room_data': g.data})
    db: Database = g.db
    try:
        if not g.data.get('name'):
            all_rooms = db.rooms.find_many() or []
            room_ids = [room.id or 0 for room in all_rooms]
            max_id = max(room_ids or [0])
            name = f"Room {max_id + 1}"
        else:
            name = g.data['name']
        room = Room(name=name)
    except ValidationError as e:
        error_message = "Error parsing room data"
        logger.error(error_message, {"error": e.errors()})
        return jsonify({"error": error_message, "details": e.errors()}), 400
    db.rooms.create(room)
    success_message = "Room created"
    model_data = room.model_dump()
    logger.info(success_message, {"data": model_data})
    return jsonify({"message": success_message, "data": model_data}), 201


@rooms_bp.get('/<int:room_id>')
@ensure_not_none('room')
def get_room(room_id):
    room = g.room
    include = request.args.get('include') or ""
    logger.info('processing get room request', {'room_id': room_id, 'include': include})
    if 'led_strip_devices' in include:
        room.devices = [device.model_dump() for device in g.db.led_strips.find_many_devices(room_id) or []]
    elif 'led_strips' in include:
        room.led_strips = [led_strip.model_dump() for led_strip in g.db.led_strips.find_many(room_id) or []]
    if 'devices' in include:
        room.devices = [device.model_dump() for device in g.db.devices.find_many(room_id) or []]
    room_data = room.model_dump()
    logger.debug('room data', {"data": room_data})
    return jsonify({"data": {"room": room_data}})


@rooms_bp.put('/<int:room_id>')
@ensure_not_none('room')
@data_has('name')
def update_room(room_id):
    logger.info('processing update room request', {'room_id': room_id, 'room_data': g.data})
    room = g.room
    room.name = g.data['name']
    db: Database = g.db
    db.rooms.update(room)
    success_message = "Room updated"
    model_data = room.model_dump()
    logger.info(success_message, {"data": model_data})
    return jsonify({"message": success_message, "data": model_data}), 200


@rooms_bp.delete('/<int:room_id>')
@ensure_not_none('room')
def delete_room(room_id):
    logger.info('processing delete room request', {'room_id': room_id})
    db: Database = g.db
    db.rooms.delete(room_id)
    success_message = "Room deleted"
    logger.info(success_message)
    return jsonify({"message": success_message}), 200


@rooms_bp.put('/<int:room_id>/led_strips')
@data_has('red', optional=True)
@data_has('green', optional=True)
@data_has('blue', optional=True)
@data_has('brightness', optional=True)
@data_has('on', optional=True)
@ensure_not_none('room')
def update_led_strips(room_id):
    logger.info('processing update led strips request', {'room_id': room_id, 'data': g.data})
    room = g.room
    db: Database = g.db
    devices = db.led_strips.find_many_devices(room_id)
    for device in devices:
        led_strip = device.led_strip
        logger.debug("led strip before update", {"led_strip": led_strip.model_dump()})
        if led_strip is None:
            raise ValueError(f"Device {device.id} has no led strip")
        if g.data.get('red') is not None:
            led_strip.red = g.data['red']
        if g.data.get('green') is not None:
            led_strip.green = g.data['green']
        if g.data.get('blue') is not None:
            led_strip.blue = g.data['blue']
        if g.data.get('brightness') is not None:
            led_strip.brightness = g.data['brightness']
        if g.data.get('on') is not None:
            logger.debug('on is not none', {'on': g.data['on']})
            led_strip.on = g.data['on']
        logger.debug("led strip after update", {"led_strip": led_strip.model_dump()})
        db.led_strips.update(led_strip)
    success_message = "Led strips updated"
    logger.info(success_message)
    return jsonify({"data": g.data}), 200

File: golang-server/api/rooms.go
package api

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/topper3418/led-hub/golang-server/db"
	"github.com/topper3418/led-hub/golang-server/models"
)

func RegisterRooms(r *gin.RouterGroup, dbPath string) {
	rooms := r.Group("/rooms")
	rooms.Use(DBMiddleware(dbPath), LoadRoom())

	rooms.GET("/", listRooms)
	rooms.POST("/", DataHas("name", true), createRoom)
	rooms.GET("/:room_id", EnsureNotNil("room"), getRoom)
	rooms.PUT("/:room_id", EnsureNotNil("room"), DataHas("name", false), updateRoom)
	rooms.DELETE("/:room_id", EnsureNotNil("room"), deleteRoom)
	rooms.PUT("/:room_id/led_strips",
		DataHas("red", true), DataHas("green", true), DataHas("blue", true),
		DataHas("brightness", true), DataHas("on", true), EnsureNotNil("room"),
		updateLedStripsInRoom,
	)
}

func listRooms(c *gin.Context) {
	database := c.MustGet("db").(*db.Database)
	rooms, _ := database.Rooms.FindMany()
	c.JSON(http.StatusOK, gin.H{"data": gin.H{"rooms": toSliceMap(rooms)}})
}

func createRoom(c *gin.Context) {
	database := c.MustGet("db").(*db.Database)
	nameVal, _ := c.Get("name")
	name := ""
	if n, ok := nameVal.(string); ok && n != "" {
		name = n
	} else {
		rooms, _ := database.Rooms.FindMany()
		maxID := 0
		for _, r := range rooms {
			if r.ID > maxID {
				maxID = r.ID
			}
		}
		name = "Room " + strconv.Itoa(maxID+1)
	}
	room := &models.Room{Name: name}
	if err := database.Rooms.Create(room); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "create failed"})
		return
	}
	c.JSON(http.StatusCreated, gin.H{
		"message": "Room created",
		"data":    structToMap(room),
	})
}

func getRoom(c *gin.Context) {
	room := c.MustGet("room").(*models.Room)
	include := strings.Split(c.Query("include"), ",")
	for i := range include {
		include[i] = strings.TrimSpace(include[i])
	}
	database := c.MustGet("db").(*db.Database)
	response := structToMap(room)
	if contains(include, "led_strip_devices") {
		devs, _ := database.LedStrips.FindManyDevices(&room.ID, false)
		response["devices"] = toSliceMap(devs)
	} else if contains(include, "led_strips") {
		strips, _ := database.LedStrips.FindMany(&room.ID)
		response["led_strips"] = toSliceMap(strips)
	}
	if contains(include, "devices") {
		devs, _ := database.Devices.FindMany(&room.ID, nil)
		response["devices"] = toSliceMap(devs)
	}
	c.JSON(http.StatusOK, gin.H{"data": gin.H{"room": response}})
}

func updateRoom(c *gin.Context) {
	room := c.MustGet("room").(*models.Room)
	nameVal, _ := c.Get("name")
	if name, ok := nameVal.(string); ok {
		room.Name = name
	}
	database := c.MustGet("db").(*db.Database)
	database.Rooms.Update(room)
	c.JSON(http.StatusOK, gin.H{
		"message": "Room updated",
		"data":    structToMap(room),
	})
}

func deleteRoom(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("room_id"))
	database := c.MustGet("db").(*db.Database)
	database.Rooms.Delete(id)
	c.JSON(http.StatusOK, gin.H{"message": "Room deleted"})
}

func updateLedStripsInRoom(c *gin.Context) {
	room := c.MustGet("room").(*models.Room)
	database := c.MustGet("db").(*db.Database)
	devices, _ := database.LedStrips.FindManyDevices(&room.ID, false)
	for _, dev := range devices {
		ls := dev.LedStrip
		if ls == nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "device has no led strip"})
			return
		}
		if vVal, ok := c.Get("red"); ok {
			if v, ok := vVal.(float64); ok {
				ls.Red = int(v)
			}
		}
		if vVal, ok := c.Get("green"); ok {
			if v, ok := vVal.(float64); ok {
				ls.Green = int(v)
			}
		}
		if vVal, ok := c.Get("blue"); ok {
			if v, ok := vVal.(float64); ok {
				ls.Blue = int(v)
			}
		}
		if vVal, ok := c.Get("brightness"); ok {
			if v, ok := vVal.(float64); ok {
				ls.Brightness = int(v)
			}
		}
		if vVal, ok := c.Get("on"); ok {
			if v, ok := vVal.(bool); ok {
				ls.On = v
			}
		}
		database.LedStrips.Update(ls)
	}
	c.JSON(http.StatusOK, gin.H{"data": c.MustGet("data")})
}

func contains(arr []string, s string) bool {
	for _, v := range arr {
		if v == s {
			return true
		}
	}
	return false
}
